src/planet-a/IcoSphereTree.cpp

```diff
@@ -501,13 +557,153 @@ TriangleSideTransform IcoSphereTree::transform_to_ancestor(
 
             out.m_scale *= 0.5f;
 
-            out.m_translation
-                    = out.m_translation * 0.5f
-                    + 0.5f * (side == (tri->m_siblingIndex + 1) % 3);
+            out.m_translation = out.m_translation * 0.5f
+                              + 0.5f * (side == (pTri->m_siblingIndex + 1) % 3);
+        }
+
+        curT = pTri->m_parent;
+    }
+
+}
+
+IcoSphereTreeObserver::Handle_T IcoSphereTree::event_add(
+        std::weak_ptr<IcoSphereTreeObserver> observer)
+{
+    return m_observers.emplace(m_observers.end(), std::move(observer));
+}
+
+void IcoSphereTree::event_remove(IcoSphereTreeObserver::Handle_T observer)
+{
+    // TODO
+}
+
+void IcoSphereTree::event_notify()
+{
+    if (!m_trianglesAdded.empty())
+    {
+        for (std::weak_ptr<IcoSphereTreeObserver> ob : m_observers)
+        {
+            // TODO: get rid of expired ones
+            ob.lock()->on_ico_triangles_added(m_trianglesAdded);
```

Comment:
Bad assumption.

Entirely possible that the weak_ptr points to a pointer that died.

If you know for sure that the weak_ptr always references a valid shared_ptr, then assert that the lock succeeds. E.g.

auto const& sp = ob.lock();
assert(sp != nullptr);
sp->on_ico_triangles_added();